presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java

```diff
@@ -87,75 +63,139 @@ public CassandraClusteringPredicatesExtractor(List<CassandraColumnHandle> cluste
+                                            columnHandle.getCassandraType()));
+                                    break;
+                                case BELOW:
+                                    rangeConjuncts.add(CassandraCqlUtils.validColumnName(columnHandle.getName()) + " < "
+                                            + CassandraCqlUtils.cqlValue(toCQLCompatibleString(range.getHigh().getValue()),
+                                            columnHandle.getCassandraType()));
+                                    break;
+                                default:
+                                    throw new AssertionError("Unhandled bound: " + range.getHigh().getBound());
+                            }
                         }
-                        return ImmutableSet.of();
-                    },
-                    allOrNone -> ImmutableSet.of());
+                    }
+                }
 
-            if (!values.isEmpty()) {
-                clusteringColumnValues.add(values);
+                if (!singleValues.isEmpty()) {
+                    if (singleValues.size() == 1) {
+                        predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " = " + singleValues.get(0);
+                    }
+                    else {
+                        predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " IN ("
+                                + Joiner.on(",").join(singleValues) + ")";
+                    }
+                }
+                else if (!rangeConjuncts.isEmpty()) {
+                    predicate = Joiner.on(" AND ").join(rangeConjuncts);
+                }
+                return predicate;
+            }, discreteValues -> {
+                if (discreteValues.isWhiteList()) {
+                    ImmutableList.Builder<Object> discreteValuesList = ImmutableList.builder();
+                    for (Object discreteValue : discreteValues.getValues()) {
+                        discreteValuesList.add(CassandraCqlUtils.cqlValue(toCQLCompatibleString(discreteValue),
+                                columnHandle.getCassandraType()));
+                    }
+                    String predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " IN ("
+                            + Joiner.on(",").join(discreteValuesList.build()) + ")";
+                    return predicate;
+                }
+                return null;
+            }, allOrNone -> null);
+
+            if (predicateString != null) {
+                // Range all not supported
+                if (predicateString.equals("Not Supported")) {
+                    break;
+                }
+                // IN restriction only on last clustering column for Cassandra version < 2.2
+                if (predicateString.contains(" IN (") && cassandraVersion != null && Float.parseFloat(cassandraVersion.substring(0, 3)) < 2.2 && currentClusteringColumn++ != (clusteringColumns.size() - 1)) {
```

Comment:
What if the `predicateString` contains `IN(` (i.e., no space in between the `IN` operator and the opening bracket)? Perhaps there is a better way to check if there is an `IN` operator in the predicate string.